File: examples/dcps/Lifecycle/CS/src/LifecycleDataPublisher.cs
using System;
using System.Threading;
using System.IO;

using DDS;
using DDS.OpenSplice;

using LifecycleData;
using DDSAPIHelper;

namespace LifecycleDataPublisher
{
    class LifecycleDataPublisher
    {
        static void usage()
        {
            Console.WriteLine("*** ERROR ***");
            Console.WriteLine("*** Usage: LifeCycleDataPublisher <autodispose_flag> [-action <action_number>]");
            Console.WriteLine("***        autodispose_flag = false | true");
            Console.WriteLine("***        -action_number   = 1 | 2 | 3 | 4");
            
            Environment.Exit(-1);
        }

        static void Main(string[] args)
        {
            char choice = ' ';
            bool interactive = true;
            bool autodispose_flag = false;

            if (args.Length < 1)
                usage();
            autodispose_flag = Boolean.Parse(args[0]);

            if (args.Length > 1)
            {
                if (!(args[1].Equals("-action")))
                    usage();
                if (args.Length < 2)
                    usage();
                choice = Char.Parse(args[2]);
                if ((!choice.Equals('1')) && (!choice.Equals('2')) && (!choice.Equals('3')) && (!choice.Equals('4')))
                    usage();
                interactive = false;
                Console.WriteLine("Choice : {0}", choice);
            }

            // DDS Helper class instances for WriterState & Msg Topic
            DDSEntityManager mgrWs = new DDSEntityManager();
            DDSEntityManager mgr = new DDSEntityManager();            

            // Create domain participant
            String partitionName = "Lifecycle";
            mgrWs.createParticipant(partitionName);

            // Create Type
            WriterStateTypeSupport wst = new WriterStateTypeSupport();
            mgrWs.registerType(wst);

            // Create Topic
            String wsTopicName = "WriterState_Msg";
            mgrWs.createTopic(wsTopicName, "Lifecycle");

            // Create Publisher
            mgrWs.createPublisher();

            // Create DataWriter
            mgrWs.createWriter();

            // Publish Samples
            IDataWriter dws_writer = mgrWs.getWriter();
            WriterStateDataWriter WriterStateWriter = dws_writer as WriterStateDataWriter;

            //-------------------Msg Topic-------------------------//
            
            // Set AutoDispose Flag
            mgr.setAutoDispose(autodispose_flag);

            // Create Participant
            mgr.createParticipant(partitionName);

            // Create type
            MsgTypeSupport mt = new MsgTypeSupport();
            mgr.registerType(mt);

            // Create Topic
            String topicName = "Lifecycle_Msg";
            mgr.createTopic(topicName, "Lifecycle");

            // Create Publisher
            mgr.createPublisher();

            // Create DataWriter
            mgr.createWriter();

            // Publish Samples
            IDataWriter dwriter = mgr.getWriter();
            MsgDataWriter LifecycleWriter = dwriter as MsgDataWriter;

            // Send MsgDataWriter state
            WriterState wsInstance = new WriterState();
            wsInstance.state = "SENDING SAMPLE";
            Console.WriteLine("===[Publisher] : ");
            Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
            ReturnCode status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
            ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");

            // Send Msg (topic to monitor)
            Msg msgInstance = new Msg();
            msgInstance.userID = 1;
            msgInstance.message = "Hello";
            Console.WriteLine("=== [Publisher]  :");
            Console.WriteLine("    userID   : {0}", msgInstance.userID);
            Console.WriteLine("    Message  : \" {0}", msgInstance.message);
            status = LifecycleWriter.Write(msgInstance, InstanceHandle.Nil);
            ErrorHandler.checkStatus(status, "MsDataWriter.Write");

            Thread.Sleep(2000);

            // Send MsgDataWriter state
            wsInstance.state = "SAMPLE SENT";
            Console.WriteLine("===[Publisher] : ");
            Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
            status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
            ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");

            if (interactive)
            {
                while ((choice != '1') && (choice != '2') && (choice != '3') && (choice != '4'))
                {
                    Console.WriteLine("=== Choose a character to continue : ");
                    Console.WriteLine("    1  Dispose the instance");
                    Console.WriteLine("    2  Unregister the instance");
                    Console.WriteLine("    3  Stop the Publisher");
                    Console.WriteLine("    4  Stop the Subscriber");
                    choice = Char.Parse(Console.In.ReadLine());
                }
            }

            switch (choice)
            {
                case '1' :
                    {
                        // Dispose instance
                        status = LifecycleWriter.Dispose(msgInstance, InstanceHandle.Nil);
                        ErrorHandler.checkStatus(status, "MsDataWriter.Dispose");
                        // Send MsgDataWriter state
                        wsInstance.state = "INSTANCE DISPOSED";
                        Console.WriteLine("===[Publisher] : ");
                        Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
                        status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
                        ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");

                        if (interactive)
                        {
                            Console.Write("\n=== enter a character to continue : ");
                            choice = Char.Parse(Console.In.ReadLine());
                        }
                        break;
                    }
                case '2':
                    {
                        // Unregister instance : the autoDispose_flag is currently
                        // ignored and the instance is never disposed automatically
                        status = LifecycleWriter.UnregisterInstance(msgInstance, InstanceHandle.Nil);
                        ErrorHandler.checkStatus(status, "MsgDataWriter.UnregisterInstance");

                        // send MsgDataWriter state
                        wsInstance.state = "INSTANCE UNREGISTERED";
                        Console.WriteLine("===[Publisher] : ");
                        Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
                        status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
                        ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");

                        if (interactive)
                        {
                            Console.Write("\n=== enter a character to continue : ");
                            choice = Char.Parse(Console.In.ReadLine());
                        }
                        break;
                    }
                case '3':
                    {
                        break;
                    }
                case '4':
                    {
                        // send MsgDataWriter state
                        wsInstance.state = "STOPPING SUBSCRIBER";
                        Console.WriteLine("===[Publisher] : ");
                        Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
                        status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
                        ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");
                        break;
                    }
            }

            // Clean-up entities for Msg topic
            // Remove the DataWriters
            mgr.getPublisher().DeleteDataWriter(LifecycleWriter);

            // send MsgDataWriter state
            wsInstance.state = "DATAWRITER DELETED";
            Console.WriteLine("===[Publisher] : ");
            Console.WriteLine("   Writer state: \" {0}", wsInstance.state, "\"");
            status = WriterStateWriter.Write(wsInstance, InstanceHandle.Nil);
            ErrorHandler.checkStatus(status, "WriterStateDataWriter.Write");

            // Remove the Publisher
            mgr.deletePublisher();

            // Remove the Topic
            mgr.deleteTopic();

            // Remove the Participant
            mgr.deleteParticipant();

            // Clean-up entities fpr WriterState topic
            // Dispose the instance
            status = WriterStateWriter.Dispose(wsInstance, InstanceHandle.Nil);
            ErrorHandler.checkStatus(status, "WriterStateWriter.Dispose");

            // Remove the DataWriter
            mgrWs.getPublisher().DeleteDataWriter(WriterStateWriter);

            // Remove the Publisher
            mgrWs.deletePublisher();

            // Remove the Topic
            mgrWs.deleteTopic();

            // Remove Participant
            mgrWs.deleteParticipant();
        }
    }
}

File: examples/dcps/Lifecycle/java5/LifecycleDataPublisher.java
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

import java.util.Collection;
import java.util.HashSet;

import org.omg.dds.core.InstanceHandle;
import org.omg.dds.core.ServiceEnvironment;
import org.omg.dds.core.policy.Durability;
import org.omg.dds.core.policy.Partition;
import org.omg.dds.core.policy.PolicyFactory;
import org.omg.dds.core.policy.Reliability;
import org.omg.dds.core.policy.WriterDataLifecycle;
import org.omg.dds.core.status.Status;
import org.omg.dds.domain.DomainParticipant;
import org.omg.dds.domain.DomainParticipantFactory;
import org.omg.dds.pub.DataWriter;
import org.omg.dds.pub.DataWriterQos;
import org.omg.dds.pub.Publisher;
import org.omg.dds.pub.PublisherQos;
import org.omg.dds.topic.Topic;
import org.omg.dds.topic.TopicQos;

import LifecycleData.Msg;

public class LifecycleDataPublisher {

	public static void usage() {
		System.out.println("*** ERROR");
		System.out.println("*** usage : LifecyclePublisher <autodispose_flag> <writer_action>");
		System.out.println("***         . autodispose_flag = false | true");
		System.out.println("***         . dispose | unregister | stoppub");
		System.exit(0);
	}

	public static void main(String args[]) {
		System.out.println("=== args_length=" + args.length);
		if (args.length < 2) {
			usage();
		}
		if (!(args[0].equals("false")) && !(args[0].equals("true"))
                    && !(args[1].equals("dispose")) && !(args[1].equals("unregister")) && !(args[1].equals("stoppub")))
		{
		  usage();
		}
		boolean autodispose = (args[0].equals("true"));
		String partitionName = "Lifecycle example";
		try {
            System.setProperty(
                      ServiceEnvironment.IMPLEMENTATION_CLASS_NAME_PROPERTY,
                      "org.opensplice.dds.core.OsplServiceEnvironment");
            ServiceEnvironment env = ServiceEnvironment
                      .createInstance(LifecycleDataPublisher.class.getClassLoader());

            DomainParticipantFactory dpf = DomainParticipantFactory.getInstance(env);

            // create Domain Participant
            DomainParticipant participant = dpf.createParticipant();

            // create Topic with reliable and transient qos
            PolicyFactory policyFactory = env.getSPI().getPolicyFactory();
            Reliability reliable = policyFactory.Reliability().withReliable();
            Durability durability = policyFactory.Durability().withTransient();
            TopicQos topicQos = participant.getDefaultTopicQos().withPolicies(reliable,durability);
            Collection<Class<? extends Status>> status = new HashSet<Class<? extends Status>>();
            Topic<Msg> topic = participant.createTopic("Lifecycle_Msg", Msg.class,topicQos, null, status);

            // create Publisher with partition qos
            Partition partition = policyFactory.Partition().withName(partitionName);
            PublisherQos pubQos = participant.getDefaultPublisherQos().withPolicy(partition);
            Publisher pub = participant.createPublisher(pubQos);

            // create DataWriter with autDisposeUnregisteredInstances qos
            WriterDataLifecycle wdlq = policyFactory.WriterDataLifecycle().withAutDisposeUnregisteredInstances(autodispose);
            DataWriterQos dwQos = pub.copyFromTopicQos(pub.getDefaultDataWriterQos().withPolicy(wdlq),topic.getQos());
            DataWriter<Msg> writer = pub.createDataWriter(topic,dwQos);
            DataWriter<Msg> writerStopper = pub.createDataWriter(topic,dwQos);

    		try {
    			Thread.sleep(500);
    		} catch (InterruptedException e) {
    			e.printStackTrace();
    		}


    		if (args[1].equals("dispose")){
    		    // Send Msg (topic to monitor)
    		    Msg msgInstance = new Msg();
    		    msgInstance.userID = 1;
    		    msgInstance.message = "Lifecycle_1";
    		    msgInstance.writerStates = "SAMPLE_SENT -> INSTANCE_DISPOSED -> DATAWRITER_DELETED";
    		    System.out.println("=== [Publisher]  :");
    		    System.out.println("    userID  : " + msgInstance.userID);
    		    System.out.println("    Message  : " + msgInstance.message);
    		    System.out.println("    writerStates : " + msgInstance.writerStates);
    		    writer.write(msgInstance, InstanceHandle.nilHandle(env));
    		    try {
    			      Thread.sleep(500);
    			} catch (InterruptedException e) {
    			      e.printStackTrace();
    			}
    			System.out.println("=== [Publisher]  : SAMPLE_SENT");

    			// Dispose instance
    			writer.dispose(InstanceHandle.nilHandle(env), msgInstance);
    		    System.out.println("=== [Publisher]  : INSTANCE_DISPOSED");
    		}
    		else if (args[1].equals("unregister")){
    		    // Send Msg (topic to monitor)
    		    Msg msgInstance = new Msg();
    		    msgInstance.userID = 2;
    		    msgInstance.message = "Lifecycle_2";
    		    msgInstance.writerStates = "SAMPLE_SENT -> INSTANCE_UNREGISTERED -> DATAWRITER_DELETED";
    		    System.out.println("=== [Publisher]  :");
    		    System.out.println("    userID  : " + msgInstance.userID);
    		    System.out.println("    Message  : " + msgInstance.message);
    		    System.out.println("    writerStates : " + msgInstance.writerStates);
    		    writer.write(msgInstance, InstanceHandle.nilHandle(env));
    		    try {
    			      Thread.sleep(500);
    			} catch (InterruptedException e) {
    			      e.printStackTrace();
    			}
    		    System.out.println("=== [Publisher]  : SAMPLE_SENT");
    		    // Unregister instance : the auto_dispose_unregisterd_instances flag
    		    // is currently ignored and the instance is never disposed
    		    // automatically
    		    writer.unregisterInstance(InstanceHandle.nilHandle(env), msgInstance);
    		    System.out.println("=== [Publisher]  : INSTANCE_UNREGISTERED");
    		}
    		else if (args[1].equals("stoppub")){
    		    // Send Msg (topic to monitor)
    		    Msg msgInstance = new Msg();
    		    msgInstance.userID = 3;
    		    msgInstance.message = "Lifecycle_3";
    		    msgInstance.writerStates = "SAMPLE_SENT -> DATAWRITER_DELETED";
    		    System.out.println("=== [Publisher]  :");
    		    System.out.println("    userID  : " + msgInstance.userID);
    		    System.out.println("    Message  : " + msgInstance.message);
    		    System.out.println("    writerStates : " + msgInstance.writerStates);
    		    writer.write(msgInstance, InstanceHandle.nilHandle(env));
    		    try {
    			      Thread.sleep(500);
    			} catch (InterruptedException e) {
    			      e.printStackTrace();
    			}
    		    System.out.println("=== [Publisher]  : SAMPLE_SENT");
    		}

    		// let the subscriber treat the previous writer state !!!!
    		System.out
    				.println("=== [Publisher] waiting 500ms to let the subscriber treat the previous write state ...");
    		try {
    			Thread.sleep(500);
    		} catch (InterruptedException e) {
    			e.printStackTrace();
    		}

    		// Clean-up the dataWriter
    		writer.close();
    		try {
    			Thread.sleep(500);
    		} catch (InterruptedException e) {
    			e.printStackTrace();
    		}
    		System.out.println("=== [Publisher]  : DATAWRITER_DELETED");

    		// Stop the subscriber
    		Msg msgInstance = new Msg();
    		msgInstance.userID = 4;
    		msgInstance.message = "Lifecycle_4";
    		msgInstance.writerStates = "STOPPING_SUBSCRIBER";
    		System.out.println("=== [Publisher]  :");
    		System.out.println("    userID  : " + msgInstance.userID);
    		System.out.println("    Message  : " + msgInstance.message);
    		System.out.println("    writerStates : " + msgInstance.writerStates);
    		writerStopper.write(msgInstance, InstanceHandle.nilHandle(env));
    		try {
    		    Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
     		participant.close();

    	} catch(Exception e) {
            System.out.println("Error occured: " + e.getMessage());
            e.printStackTrace();
        }
	}
}
